### PR TITLE
Add accurate resource group gas metering

### PR DESCRIPTION
**Pull request by gerben-stavenga**
### Description

Refactor the code to return actual bytes read from storage to the gas metering. Use gas version feature to enable this when the next version is bumped.

### CODE REVIEW

**Review by movekevin — APPROVED** [review]

File: third_party/move/move-core/types/src/resolver.rs [hunk 1]
```diff
@@ -41,7 +42,7 @@ pub trait ResourceResolver {
         address: &AccountAddress,
         typ: &StructTag,
         metadata: &[Metadata],
-    ) -> Result<Option<Vec<u8>>, Error>;
+    ) -> Result<(Option<Vec<u8>>, Option<NumBytes>), Error>;
```
Comment (vgao1996) [inline third_party/move/move-core/types/src/resolver.rs:45]: Should make this `Option<(Vec<u8>, NumBytes)>` instead.

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 2]
```diff
@@ -65,32 +76,40 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
         address: &AccountAddress,
         struct_tag: &StructTag,
         metadata: &[Metadata],
-    ) -> Result<Option<Vec<u8>>, VMError> {
+    ) -> Result<(Option<Vec<u8>>, Option<NumBytes>), VMError> {
```
Comment (vgao1996) [inline aptos-move/aptos-vm/src/data_cache.rs:79]: Same here. Should merge the two options.

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 3]
```diff
@@ -42,16 +47,22 @@ pub(crate) fn get_resource_group_from_metadata(
 /// Adapter to convert a `StateView` into a `MoveResolverExt`.
 pub struct StorageAdapter<'a, S> {
     state_store: &'a S,
+    accurate_byte_count: Cell<bool>,
```
Comment (vgao1996) [inline aptos-move/aptos-vm/src/data_cache.rs:50]: Why does this need to be a cell? It's set only during creation time, right?

**Review by vgao1996 — COMMENTED** [review]
Some early feedback

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 4]
```diff
@@ -65,32 +76,40 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
...
+                let mut len = Some(NumBytes::new(group_data.len() as u64));
                 let group_data: BTreeMap<StructTag, Vec<u8>> = bcs::from_bytes(&group_data)
                     .map_err(|_| {
                         PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
                             .finish(Location::Undefined)
                     })?;
                 let res = group_data.get(struct_tag).cloned();
+                if !self.accurate_byte_count.get() {
+                    len = res.as_ref().map(|b| NumBytes::new(b.len() as u64));
+                }
```
Comment (vgao1996) [inline aptos-move/aptos-vm/src/data_cache.rs:102]: Yeah, as we discussed offline, let's try to make the amount charged independent of the ordering by which you load the resources in the same group.

**Comment by vgao1996** [comment]
Also a reminder that we need to take care of the lifetime issue

**Comment by gerben-stavenga** [comment]
> Also a reminder that we need to take care of the lifetime issue we just discussed

I addressed the issue by releasing the cache after proloque. This was infrastructure that was already there for assembling resource groups for creating the writeset

File: aptos-move/aptos-vm/src/data_cache.rs [hunk 5]
```diff
@@ -42,16 +42,22 @@ pub(crate) fn get_resource_group_from_metadata(
...
 impl<'a, S: StateView> StorageAdapter<'a, S> {
     pub fn new(state_store: &'a S) -> Self {
-        Self {
+        let mut s = Self {
             state_store,
+            accurate_byte_count: false,
             resource_group_cache: RefCell::new(BTreeMap::new()),
-        }
+        };
+        if gas_config(&s).1 >= 9 {
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/data_cache.rs:57]: you need to set the version to 9 in this PR too otherwise during upgrade it won't be set.

**Review by vgao1996 — APPROVED** [review]
@gerben-stavenga I think David made a good point about bumping LATEST_GAS_FEATURE_VERSION. Please do that before you land this!

**Comment by sherry-x** [comment]
Triggering re-run failed job to confirm the performance numbers.

**Pull request merged**